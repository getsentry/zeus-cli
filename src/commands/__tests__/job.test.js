'use strict';

/* eslint-env jest */

const commandUpdate = require('../job_cmds/update');
const getEnv = require('../../environment');
const logger = require('../../logger');
const Zeus = require('../../sdk');

jest.mock('../../environment');
jest.mock('../../logger');
jest.mock('../../sdk/client');

describe('job add command', () => {
  let func;

  beforeEach(() => {
    jest.clearAllMocks();
    Zeus.instance.createOrUpdateBuild = jest.fn();
    Zeus.instance.createOrUpdateBuild.mockImplementation(() =>
      Promise.resolve({})
    );
    Zeus.instance.createOrUpdateJob = jest.fn();
    Zeus.instance.createOrUpdateJob.mockImplementation(() =>
      Promise.resolve({})
    );
    func = Zeus.instance.createOrUpdateJob;
    getEnv.mockReturnValue({});
  });

  test('adds a new job', () => {
    const argv = {
      build: 1,
      'build-label': 'New job',
      url: 'https://invalid/job/1',
    };

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the job ID from the environment', () => {
    getEnv.mockReturnValue({ jobId: 12345 });
    const argv = {};

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the build ID from the environment', () => {
    getEnv.mockReturnValue({ buildId: 2345 });
    const argv = {};

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the commit ID from the environment', () => {
    getEnv.mockReturnValue({ commitId: '1234567' });
    const argv = {};

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the build label from the environment', () => {
    getEnv.mockReturnValue({ buildLabel: 'new build label' });
    const argv = {};

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('uses the job label from the environment', () => {
    getEnv.mockReturnValue({ jobLabel: 'new job label' });
    const argv = {};

    expect.assertions(1);
    return commandUpdate.handler(argv).then(() => {
      expect(func.mock.calls[0]).toMatchSnapshot();
    });
  });

  test('logs the successful job update', () => {
    expect.assertions(1);
    return commandUpdate.handler({}).then(() => {
      const successMessage = 'Job updated';
      expect(logger).toHaveBeenCalledWith(
        expect.stringMatching(successMessage)
      );
    });
  });

  test('logs errors from the SDK', () => {
    func.mockImplementation(() =>
      Promise.reject(new Error('expected failure'))
    );

    expect.assertions(1);
    return commandUpdate.handler({}).then(() => {
      expect(logger).toHaveBeenCalledWith(
        expect.stringMatching(/expected failure/)
      );
    });
  });
});
