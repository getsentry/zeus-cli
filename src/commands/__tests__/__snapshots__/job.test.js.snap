// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`job add command adds a new job 1`] = `
Array [
  Object {
    "build": 1,
    "buildLabel": "New job",
    "job": undefined,
    "jobLabel": undefined,
    "ref": undefined,
    "status": undefined,
    "url": "https://invalid/job/1",
  },
]
`;

exports[`job add command uses the build ID from the environment 1`] = `
Array [
  Object {
    "build": 2345,
    "buildLabel": undefined,
    "job": undefined,
    "jobLabel": undefined,
    "ref": undefined,
    "status": undefined,
    "url": undefined,
  },
]
`;

exports[`job add command uses the build label from the environment 1`] = `
Array [
  Object {
    "build": undefined,
    "buildLabel": "new build label",
    "job": undefined,
    "jobLabel": undefined,
    "ref": undefined,
    "status": undefined,
    "url": undefined,
  },
]
`;

exports[`job add command uses the commit ID from the environment 1`] = `
Array [
  Object {
    "build": undefined,
    "buildLabel": undefined,
    "job": undefined,
    "jobLabel": undefined,
    "ref": "1234567",
    "status": undefined,
    "url": undefined,
  },
]
`;

exports[`job add command uses the job ID from the environment 1`] = `
Array [
  Object {
    "build": undefined,
    "buildLabel": undefined,
    "job": 12345,
    "jobLabel": undefined,
    "ref": undefined,
    "status": undefined,
    "url": undefined,
  },
]
`;

exports[`job add command uses the job label from the environment 1`] = `
Array [
  Object {
    "build": undefined,
    "buildLabel": undefined,
    "job": undefined,
    "jobLabel": "new job label",
    "ref": undefined,
    "status": undefined,
    "url": undefined,
  },
]
`;
